cartoframes/io/context.py

```diff
@@ -0,0 +1,329 @@
+
+        df = pd.read_csv(
+            raw_result,
+            converters=converters,
+            parse_dates=parse_dates)
+
+        if index_col:
+            df.index = df[index_col]
+            df.index.name = None
+
+        return df
+
+    def _copy_from(self, dataframe, table_name, dataframe_columns_info):
+        query = """
+            COPY {table_name}({columns}) FROM stdin WITH (FORMAT csv, DELIMITER '|', NULL '{null}');
+        """.format(
+            table_name=table_name, null=PG_NULL,
+            columns=','.join(c.database for c in dataframe_columns_info.columns)).strip()
+
+        data = _rows(dataframe, dataframe_columns_info)
+
+        self.copy_client.copyfrom(query.strip(), data)
+
+    def _normalize_table_name(self, table_name):
+        norm_table_name = normalize_name(table_name)
+        if norm_table_name != table_name:
+            print('Debug: table name normalized: "{}"'.format(norm_table_name))
+        return norm_table_name
+
+
+def _drop_table_query(table_name, if_exists=True):
+    return '''DROP TABLE {if_exists} {table_name}'''.format(
+        table_name=table_name,
+        if_exists='IF EXISTS' if if_exists else '')
+
+
+def _create_table_query(table_name, columns):
+    cols = ['{column} {type}'.format(column=c.database, type=c.database_type) for c in columns]
+
+    return '''CREATE TABLE {table_name} ({cols})'''.format(
+        table_name=table_name,
+        cols=', '.join(cols))
+
+
+def _cartodbfy_query(table_name, schema):
+    return "SELECT CDB_CartodbfyTable('{schema}', '{table_name}')" \
+        .format(schema=schema, table_name=table_name)
+
+
+def _create_auth_client(credentials, public=False):
+    return APIKeyAuthClient(
+        base_url=credentials.base_url,
+        api_key='default_public' if public else credentials.api_key,
+        session=credentials.session,
+        client_id='cartoframes_{}'.format(__version__),
+        user_agent='cartoframes_{}'.format(__version__)
+    )
+
+
+def _rows(df, dataframe_columns_info):
+    for index, _ in df.iterrows():
+        row_data = []
+        for c in dataframe_columns_info.columns:
+            col = c.dataframe
+            if col not in df.columns:
+                if df.index.name and col == df.index.name:
+                    val = index
+                else:  # we could have filtered columns in the df. See DataframeColumnsInfo
+                    continue
+            else:
+                val = df.at[index, col]
+
+            if dataframe_columns_info.geom_column and col == dataframe_columns_info.geom_column:
+                geom = decode_geometry(val, dataframe_columns_info.enc_type)
+                if geom:
+                    val = 'SRID=4326;{}'.format(geom.wkt)
+                else:
+                    val = ''
+
+            row_data.append(encode_row(val))
+
+        csv_row = b'|'.join(row_data)
+        csv_row += b'\n'
+
+        yield csv_row
+
+
+# def _rename_index_for_upload(self):
```

Comment:
This was used in geocoding stuff. We have a bug related to that. Are you removing it? If I recall well, the idea of that was to have an id to merge between the original and geocoded data.

Reply:
Yes. This is a reminder to use it. I'll let you know when it's used.

Reply:
This is the issue: https://github.com/CartoDB/cartoframes/issues/1211 But we can tackle it after this one